### DISCUSSION TITLE
How to change BackendUrl after static generation

### DISCUSSION THREAD

**Discussion by skoenfaelt** [nuxt/nuxt#10701] [Questions (Nuxt 2)] [answered]
## Current Project Architecture
- Nuxt 2 App (v2.15.8)
- ssr is set on true (default)
- built with target static
- using runtimeConfigs for dynamic deployment with different URLs of the Backend
- CI/CD create server-specific .env and "generates" the app via `nuxt generate`

## The Goal
I got the task, we should be able to change the **Backend Url** after the app is built.

This is because our PM would be able to copy the Server Deployed Master and use it for presentations on a separate demo instance (or on dynamic customer side presentations) therefore, _**he won't want to trigger a new deployment**_, "just to change the API URL".

### Use configurable JSON in `assets` folder
- this wasn't a good idea, since webpack puts it into js chunks > so, not easy editable for the PM

> *Could I exclude a File from Webpack to bundle it?*, did not find a working solution

### Use configurable JSON in `static` folder
- easy to configure for the PM
- but at build timeJSONxt reads the json fiea lds, but it into store or call the file inside a plugin, (I thinJSONxt knows the json is in the static folder and therefore it assumes, this file will not change) > so it put the value of the JSON inside json in every component, so I can't change it anymore
- I can change the JSON inside the json, but nothing changes in the code, since the build step "hardcoded" it

### using .env on the server
As far as I know, the static files do not have access to the env variables, and we do not run a nodeJs server on our app hosting server.

---

So currently I have no Idea to solve this NuxtJs is a nice Framework and I use it for many years, but this problem gives me a headache.

Just to mention, our PM argument is, "in our other app,..." (which is built with angular) "... it is 'easy' possible"

He is right, in this app, we have a JSON, which is loaded in runtime. So is there a way to build this in nuxt?

**Comment by ghost** [comment]
Do you mean by "backend api" nuxt's sever-middleware.
If so you can set the path the middleware with your env-Variable. With Nuxt env-variables are directly exposed to nuxt.config.js and code run with node, like plugins/modules etc. like this for example:
process.env.CUSTOM_MIDDLEWARE_PATH

Which you could assign the path of your middleware

{ path: process.env.CUSTOM_MIDDLEWARE_PATH, handler: '~/server-middleware/index.js' }, 

https://nuxtjs.org/docs/configuration-glossary/configuration-servermiddleware/#usage

If you correctly configured your CI/CD Pipeline and nuxt has access to your set env-variables, this should work without a hitch

**Reply by skoenfaelt** [reply]
Our Backend is an ASP .Net Core backend which is often hosted on a different server. Of course for our testsuite we can handle this via .env files.

Some of our deployment steps are a bit old, and for those deployments we do not always know the address of the BE server. In those cases the server admin should write the URL of the BE API into the config file. I think static hosting is simply not made for this use case.

**Reply by ghost** [reply]
I also don't think you can just change files while being statically hosted by nuxt. That said, depending on where you host the nuxt app, you could write either a small server-middleware or serverless function (eg. Netlify or Vercel) that actually reads that json-file on request and then calls the backend-api based on its content.

What I still don't get is: How someone could directly edit files, after the nuxt app is deployed. And why you'd want that.

**Reply by ghost** [reply]
In my opinion it would still be better to use .env variables for such a use case. You can expose env vars via a runtime config object. publicRuntimeConfig in nuxt.config.js exposes it's values to server but also client-side code.
https://nuxtjs.org/docs/configuration-glossary/configuration-runtime-config/

If the env-variable should be changed, the app need to be rebuild and redeployed to actually apply the change.

**Comment by skoenfaelt** [comment, marked as answer]
Closing this, since the use case of my Project manager is not state of the art... If he wants that customizability we will use the node hosting. This also solves the problem with the dynamic pages. We won't have a `getAllUsers` endpoint. So we wont get pages like /user/id/someOtherRoute/SomeOtherID

In the end its not solved but we simply overuse the static feature
